crates/llms/src/bedrock/chat/util.rs

```diff
@@ -0,0 +1,284 @@
+}
+
+pub(super) fn tool_config(
+    tools: Option<Vec<ChatCompletionTool>>,
+    tool_choice: Option<ChatCompletionToolChoiceOption>,
+) -> Option<ToolConfiguration> {
+    let tool_choice = match tool_choice {
+        Some(ChatCompletionToolChoiceOption::Auto) => {
+            Some(ToolChoice::Auto(AutoToolChoiceBuilder::default().build()))
+        }
+        Some(ChatCompletionToolChoiceOption::Required) => {
+            Some(ToolChoice::Any(AnyToolChoiceBuilder::default().build()))
+        }
+        Some(ChatCompletionToolChoiceOption::Named(ChatCompletionNamedToolChoice {
+            function: FunctionName { name },
+            ..
+        })) => SpecificToolChoice::builder()
+            .name(name)
+            .build()
+            .ok()
+            .map(ToolChoice::Tool),
+        _ => None, // None, ChatCompletionToolChoiceOption::None, or Unknown.
+    };
+    let tools = tools?
+        .into_iter()
+        .filter_map(|t| {
+            let FunctionObject {
+                name,
+                description,
+                parameters,
+                ..
+            } = t.function;
+            Some(Tool::ToolSpec(
+                ToolSpecification::builder()
+                    .name(name)
+                    .set_description(description)
+                    .set_input_schema(
+                        parameters.map(|p| ToolInputSchema::Json(value_to_document(p))),
+                    )
+                    .build()
+                    .ok()?,
+            ))
+        })
+        .collect::<Vec<_>>();
+
+    ToolConfiguration::builder()
+        .set_tool_choice(tool_choice)
+        .set_tools(Some(tools))
+        .build()
+        .ok()
+}
+
+#[allow(deprecated)]
+pub(super) fn chat_choice_stream(
+    content: Option<String>,
+    tool_calls: Option<Vec<ChatCompletionMessageToolCallChunk>>,
+    role: Option<Role>,
+    refusal: Option<String>,
+    finish_reason: Option<FinishReason>,
+) -> ChatChoiceStream {
+    ChatChoiceStream {
+        index: 0,
+        delta: ChatCompletionStreamResponseDelta {
+            content,
+            function_call: None,
+            tool_calls,
+            role,
+            refusal,
+        },
+        finish_reason,
+        logprobs: None,
+    }
+}
+
+#[allow(clippy::cast_sign_loss)]
+pub(super) fn convert_usage(usage: &TokenUsage) -> CompletionUsage {
+    let TokenUsage {
+        input_tokens,
+        output_tokens,
+        total_tokens,
+        cache_read_input_tokens,
+        ..
+    } = usage;
+    CompletionUsage {
+        prompt_tokens: *input_tokens as u32,
+        completion_tokens: *output_tokens as u32,
+        total_tokens: *total_tokens as u32,
+        prompt_tokens_details: cache_read_input_tokens.map(|t| PromptTokensDetails {
+            cached_tokens: Some(t as u32),
+            audio_tokens: None,
+        }),
+        completion_tokens_details: None,
+    }
+}
+
+pub(super) fn value_to_document(value: serde_json::Value) -> Document {
```

Comment:
I'm pretty sure we have very similar code for S3 vectors - might be a candidate for reuse?

Reply:
https://github.com/spiceai/spiceai/issues/6676